### ISSUE TITLE
Clone and apply checkpoint execution recipes with explicit overrides and evaluation/training scopes

### ISSUE THREAD

**Issue by amadou-6e** [amadou-6e/theseo-anysearch#473]
## Objective
Provide a general anysearch clone/apply workflow for executing a saved policy on its original or a replacement world, without silently changing the execution contract. This also supplies standalone evaluation without training.

## Proposed CLI
- anysearch clone --checkpoint <path> --scope evaluation --output evaluation.yaml
- anysearch apply evaluation.yaml --world <manifest> --dry-run
- anysearch apply evaluation.yaml --world <manifest>
- Explicit component overrides/exclusions in the recipe or clone options, e.g. --disable reward:segment_countdown_goal. Exact flag spelling is provisional.

## Execution recipe and reproducibility
A checkpoint alone is not currently a self-contained execution bundle. Resolve its original run metadata and create a versioned recipe referencing verified policy/connector artifacts, resolved environment/model settings, action meanings, observation encoders/normalization, movement/collision/termination/reward rules, extensions and capability bindings, task/route sampling, world/catalog assets, seeds, source identities and dependency/runtime versions. Record artifact hashes and resolve paths portably. Offer copying/bundling for portability while permitting references to verified immutable artifacts. Missing provenance requires explicit original inputs or a clear failure, never guesses from current defaults. Distinguish reproducible execution from bit-for-bit hardware/runtime determinism.

## Scopes
Evaluation: preserve policy weights, observations/actions/connectors and environment rules; disable exploration and weight updates; do not execute learner/optimizer machinery; preserve task selection and freeze adaptive curriculum changes. Training continuation: restore optimizer/learner and supported continuation state, original exploration and curriculum adaptation. Fine-tuning: explicitly initialize from policy weights with reset learner/optimizer semantics; never label it faithful continuation. Enumerate unsupported restoration state rather than claiming exact continuation. Scope-inactive components are distinct from intentional exclusions.

## Overrides and exclusions
Inherit execution behavior by default. Replace only geometry/world when requested, preserving every other compatible policy-facing setting. Validate action and observation spaces and semantic bindings, not only tensor dimensions. Require explicit replacement tasks/samplers when routes depend on the old geometry. Intentionally disable or replace capabilities rather than entire extension libraries: a library may combine reward, movement, predicates, outcomes and scenarios. Check dependencies and reject exclusions that break policy inputs/connectors or environment behavior. If disabling a reward requires an alternative, require an explicit named replacement or explicit supported no-reward mode; never silently select a default. Changed worlds or execution components produce a new execution identity, not an unchanged replay or historical resume.

segment_countdown_goal is training-oriented reward shaping but is currently an environment reward provider, not training-only: configured evaluation uses it too. Removing it changes returns, and may affect actions if rewards feed policy inputs/connectors or other execution logic. Audit the actual dependencies before declaring it behavior-neutral.

## Dry run and outputs
Clone only prepares the recipe; apply --dry-run performs read-only validation and reports inherited/restored/inactive/disabled/replaced settings plus unresolved dependencies. Apply executes with fresh output identity, preserves source checkpoint/run/assets, and writes resolved recipe, difference manifest, checkpoint/world/artifact hashes, seeds, runtime provenance, metrics and replayable trajectories. Evaluation reports success, collisions, return and episode length without training. Training/fine-tuning budgets must be explicit in the recipe; cloning does not start a run or grant extra compute. Configurable omissions are recorded and cannot be presented as full reproducibility.

## Acceptance
- Versioned recipe/schema and documented clone/apply CLI with evaluation, continuation and fine-tuning semantics.
- Legacy checkpoints resolve through parent run or explicit supplied provenance; failures are informative.
- Tests for same-world inherited execution, replacement-world evaluation, intentional reward replacement, multi-capability extensions, incompatible spaces/semantics, missing/moved/tampered artifacts and geometry-dependent task overrides.
- Dry-run is non-executing; cloning is non-mutating to sources; evaluation does not update weights or run training.
- Deterministic seeded equivalence checks where supported, explicit determinism limits, inspectable difference/provenance manifests and replayable output.

General reusable runtime/CLI infrastructure, not a change to any frozen perception-encoder experiment or authorization to launch training. Follow the normal issue-number branch and develop PR workflow; experiment usage must separately obey its pinned protocol and run-identity requirements.

**Comment by amadou-6e** [comment]
# Review: checkpoint clone/apply execution recipes

Checked this against the actual code rather than taking the two central factual claims at face value — both hold up.

## Both grounding claims are verified true, not just plausible

- **"A checkpoint alone is not currently a self-contained execution bundle."** Read `rllib/trainer/checkpointing.py`: `CheckpointState` stores only `iteration`, `episodes_total`, `rllib_path`, and a `world_contract`/`world_fingerprint` pair (checked on restore, raises on mismatch). That's it — no reward-provider identity, action/observation space config, movement/collision/termination rules, extension/capability bindings, task/route sampling settings, seeds beyond what the world contract captures, or dependency/runtime versions. The world-contract check is a real but narrow guard; everything else this issue wants to formalize genuinely isn't captured today.
- **"`segment_countdown_goal` is... currently an environment reward provider, not training-only: configured evaluation uses it too."** Confirmed directly in `usage/experiments/tune/waypoint_route_segment_reward/experiment.yaml`: `env.rewards.custom.name: segment_countdown_goal` is set once at the shared `env:` block level, and the `evaluation:` block only adds evaluation-specific scheduling (`episodes`, `seed`, `min_success_rate`, curriculum settings) — there's no separate evaluation-scoped reward override. Evaluation episodes genuinely run through the same reward provider as training. This is a well-chosen concrete example, not a hypothetical.

## Consistent with the infrastructure it builds on

`--world <manifest>` correctly reuses the world-manifest/compiled-pack concept from the routing contracts and world-provider work rather than inventing a parallel reference scheme, and the evaluation-scope description (preserve task selection, freeze curriculum changes) matches the `evaluation.waypoint_curriculum` structure already used elsewhere in this codebase. The hash-everything, fail-explicit-not-silently approach ("missing provenance requires explicit original inputs or a clear failure, never guesses from current defaults") matches the pattern used consistently across the routing/world-provider contracts (`routing_manifests.py`, `source_cache.py`) — this proposal is following an established, working convention rather than introducing a new one.

## One connected question worth adding to the acceptance criteria

Given the world-provider `add` command's split-leakage protection is scoped to worlds explicitly registered in a shared `study_id` (per my earlier review of that spec), does `apply ... --world <manifest>` have any awareness of a target world's split role (train/calibration/test) when evaluating a checkpoint against it? It's plausible this is intentionally out of scope (evaluating a trained checkpoint against its own training world is often exactly the point), but if a future use case is "evaluate on a held-out world," it'd be worth deciding now whether `apply` should surface or check that role rather than treating every world manifest identically.

No blockers — well-scoped, technically accurate, and appropriately explicit about the hard parts (multi-capability extension exclusion, reward-affects-actions risk, reproducibility-vs-bit-exact-determinism). Good proposal to build from.

**Comment by amadou-6e** [comment]
Implementation started on feat/473. First commit adds versioned clone recipes and read-only apply validation. It was exercised against real run 5034f160 checkpoint 100: checkpoint tree, archived experiment and segment_countdown_goal extension artifacts all verify; the historical missing source revision is surfaced as a provenance gap. Execution remains deliberately blocked. Draft PR records remaining restoration/evaluation/continuation/fine-tuning work; no run was launched and no source artifact was mutated.

**Comment by amadou-6e** [comment]
Campaign checkpoint: foundation PR #474 merged to develop as 130c110. Work split into dependency-ordered issues #482 portable bundles, #483 archived-config/policy compatibility, #484 overrides, #485 evaluation executor, #486 continuation/fine-tuning, #487 hardening. #482 is implemented in PR #488 (commit 3aca39b): bundle-relative copied artifacts, checkpoint tree/config/extension hashes, relocation/tamper validation, and computed source-revision provenance; 6 focused tests pass. PR #488 CI is running and no review exists yet. Per repository dependency policy, #483 cannot start until #488 is reviewed and merged unless the user explicitly authorizes stacking unmerged work. No policy execution or training process is active. Next action: review/merge #488, then branch #483 from updated develop.

**Comment by amadou-6e** [comment]
Campaign checkpoint 2026-09-24: #482/#488, #483/#496, and #484/#497 are merged to develop. #485 implementation is in PR #498 (5591d48), with 61 focused tests and two real checkpoint evaluation smokes; CI pending. Next dependency: integrate #498, then implement #486 continuation/fine-tuning, then #487 hardening. No active experiment process; local smoke artifacts under runtime/execution-485-smoke. Source revision missing from the historical 5034f160 checkpoint is explicitly recorded as a provenance gap.

**Comment by amadou-6e** [comment]
Campaign checkpoint: #485/#498 merged to develop at 4ae7222. #486 implementation is in PR #499 (8faedcc), with passing real PPO continuation/fine-tuning smoke and explicit refusal of incomplete/staged continuation; CI pending. Next: integrate #499, then #487. No active training process. The historical large-world checkpoint remains evaluation/fine-tuning eligible but not exact-continuation eligible because its curriculum state was not stored per checkpoint.

**Comment by amadou-6e** [comment]
Campaign checkpoint: #482/#488, #483/#496, #484/#497, #485/#498, and #486/#499 are merged to develop. Final hardening #487 is in PR #500 (3e54f5e): 38 focused tests and real Ray end-to-end passed, CI pending. Next: review/merge #500 and close #487/#473 if accepted. No active training or evaluation process. Known explicit limits: single-agent PPO; staged training checkpoints and historical per-checkpoint curriculum omissions cannot claim exact continuation.

**Comment by amadou-6e** [comment]
Final campaign result (2026-09-24): completed and integrated on develop. The recipe foundation (#474), portable artifact integrity/provenance (#488), evaluation executor (#498), continuation/fine-tuning executor (#499), and end-to-end hardening (#500, squash fd94554c9620096f94aaed95ffda6a1bbd4711ec) are merged. #500 passed all CI checks, 38 focused tests, and a real Ray end-to-end run covering moved-bundle seeded equivalence, compiled replacement-world evaluation, exact continuation, and weight-only fine-tuning. Execution writes fresh identity, resolved configuration/differences, metrics and replayable trajectories; dry-run preflights without starting Ray or writing a run. Supported scope is single-agent PPO. Exact continuation intentionally rejects historical checkpoints without checkpoint-local curriculum/early-stop snapshot, staged runs, and incompatible worlds or policy spaces; fine-tuning is explicitly new training, not a faithful resume. Hard process kills can leave partial output; caught interruptions write a failure record. No experiment training was launched as part of this infrastructure campaign. Disposition: promote; parent objective met.
